sycl/test-e2e/Matrix/element_wise_odd_sizes_bf16_impl.hpp

```diff
@@ -0,0 +1,107 @@
+public:
+  T *mat;
+
+public:
+  T *get_data() { return mat; }
+  void set_data(T *data) { mat = data; }
+  big_matrix(T *data) : mat(data) {}
+};
+
+template <typename T, size_t M, size_t N>
+void assert_ops_ref(host_accessor<T, 2, access::mode::read> C,
+                    const float ref) {
+  for (size_t i = 0; i < M; i++)
+    for (size_t j = 0; j < N; j++) {
+      auto diff = make_fp32(C[i][j]) - ref;
+      assert(std::fabs(static_cast<float>(diff)) <
+             std::numeric_limits<float>::epsilon());
+    }
+}
+template <typename T, size_t M, size_t N, size_t TM, size_t TN, size_t TK>
+void matrix_verify_add(queue q, big_matrix<T, M, N> &A, nd_range<2> &r,
+                       const float ref) {
+  buffer<bfloat16, 2> bufA(A.get_data(), range<2>(M, N));
+
+  q.submit([&](handler &cgh) {
+     auto accA = bufA.get_access<access::mode::read_write>(cgh);
+
+     cgh.parallel_for<class add_matrix>(
+         r, [accA](nd_item<2> spmd_item) [[intel::reqd_sub_group_size(SG_SZ)]] {
+           const auto global_idx = spmd_item.get_global_id(0);
+           const auto global_idy = spmd_item.get_global_id(1);
+           const auto sg_startx = global_idx - spmd_item.get_local_id(0);
+           const auto sg_starty = global_idy - spmd_item.get_local_id(1);
+
+           sub_group sg = spmd_item.get_sub_group();
+           joint_matrix<sub_group, T, use::a, TM, TK, layout::row_major> sub_a;
+
+           joint_matrix_fill(sg, sub_a, bfloat16(5.0));
+
+           auto wi_slice_a =
+               sycl::ext::intel::experimental::matrix::get_wi_data(sg, sub_a);
+           for (int i = 0; i < wi_slice_a.length(); i++) {
+             wi_slice_a[i] = wi_slice_a[i] + bfloat16(2);
+           }
+
+           ext::intel::experimental::matrix::joint_matrix_store(
+               sg, sub_a,
+               accA.get_pointer() + (sg_startx * TM) * N +
+                   sg_starty / SG_SZ * TN,
+               N);
+         }); // parallel for
+   }).wait();
+  assert_ops_ref<T, M, N>(bufA.get_host_access(read_only), ref);
+}
+
+void matrix_ops_ref(float *D, int M, int N) {
+  for (int m = 0; m < M; m++)
+    for (int n = 0; n < N; n++) {
+      *(D + m * N + n) = 0;
+      *(D + m * N + n) *= 2;
+    }
+}
+
+template <typename Ta, int vnni_factor, size_t tM, size_t tN, size_t tK>
+int add_ref() {
+
+  static constexpr size_t MATRIX_M = tM * M_MULTIPLIER;
+  static constexpr size_t MATRIX_N = 128;
+  Ta A[MATRIX_M][MATRIX_N];
+  float D[MATRIX_M][MATRIX_N];
+
+  big_matrix<float, MATRIX_M, MATRIX_N> MD((float *)&D);
+  big_matrix<Ta, MATRIX_M, MATRIX_N> MA((Ta *)&A);
+
+  size_t NDRangeM = MATRIX_M / tM;
+  size_t NDRangeN = MATRIX_N / tN;
+  queue q;
+  nd_range<2> r({NDRangeM, NDRangeN * SG_SZ}, {1, 1 * SG_SZ});
+
+  matrix_verify_add<Ta, MATRIX_M, MATRIX_N, tM, tN, tK>(q, MA, r, 7.0);
+
+  return 0;
+}
+
+int main() {
+  add_ref<bfloat16, 1 /*VF*/, 1 /*TM*/, SG_SZ /*TN*/, 16 /*TK*/>();
```

Comment:
please add the int8 cases as in https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Matrix/joint_matrix_all_sizes_impl.hpp

Reply:
Done